influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -0,0 +1,804 @@
+    fn size(&self) -> usize {
+        match self {
+            CacheEntry::Fetching => 0,
+            CacheEntry::Success(v) => v.size(),
+            CacheEntry::Failed => 0,
+            CacheEntry::Deleted => 0,
+            CacheEntry::TooLarge => 0,
+        }
+    }
+
+    fn is_fetching(&self) -> bool {
+        matches!(self, CacheEntry::Fetching)
+    }
+
+    fn is_success(&self) -> bool {
+        matches!(self, CacheEntry::Success(_))
+    }
+
+    fn keep(&self) -> bool {
+        self.is_fetching() || self.is_success()
+    }
+}
+
+/// Implements the [`WeightScale`] trait to determine a [`CacheEntry`]'s size on insertion to
+/// the cache
+#[derive(Debug)]
+struct CacheEntryScale;
+
+impl WeightScale<Path, CacheEntry> for CacheEntryScale {
+    fn weight(&self, key: &Path, value: &CacheEntry) -> usize {
+        key.as_ref().len() + value.size()
+    }
+}
+
+/// Placeholder name for formatting datafusion errors
+const STORE_NAME: &str = "mem_cached_object_store";
+
+/// An object store with an associated cache that can serve GET-style requests using the cache
+///
+/// The least-recently used (LRU) entries will be evicted when new entries are inserted, if the
+/// new entry would exceed the cache's memory capacity
+#[derive(Debug)]
+pub struct MemCachedObjectStore {
+    /// An inner object store for which items will be cached
+    inner: Arc<dyn ObjectStore>,
+    /// A weighted LRU cache for storing the objects associated with a given path in memory
+    // NOTE(trevor): this uses a mutex as the CLruCache type needs &mut self for its get method, so
+    // we always need an exclusive lock on the cache. If this creates a performance bottleneck then
+    // we will need to look for alternatives.
+    //
+    // A Tokio mutex is used to prevent blocking the thread while waiting for a lock, and so that
+    // the lock can be held accross await points.
+    cache: Arc<Mutex<CLruCache<Path, CacheEntry, RandomState, CacheEntryScale>>>,
+}
+
+impl MemCachedObjectStore {
+    /// Create a new [`MemCachedObjectStore`] with the given memory capacity
+    fn new(inner: Arc<dyn ObjectStore>, memory_capacity: usize) -> Self {
+        let cache = CLruCache::with_config(
+            CLruCacheConfig::new(NonZeroUsize::new(memory_capacity).unwrap())
+                .with_scale(CacheEntryScale),
+        );
+        Self {
+            inner,
+            cache: Arc::new(Mutex::new(cache)),
+        }
+    }
+
+    /// Get an entry in the cache if it contains a successful fetch result, or `None` otherwise
+    ///
+    /// This requires `&mut self` as the underlying method on the cache requires a mutable reference
+    /// in order to update the recency of the entry in the cache
+    async fn get_cache_value(&self, path: &Path) -> Option<Arc<CacheValue>> {
+        self.cache
+            .lock()
+            .await
+            .get(path)
+            .and_then(|entry| match entry {
+                CacheEntry::Fetching
```

Comment:
I opened https://github.com/influxdata/influxdb/issues/25383 to address this since this could be tricky. But, yeah, returning `None` here basically means that we make an additional request to the object store, which is not ideal.